### PR TITLE
pallet-collective: allow customized default vote

### PR DESCRIPTION
**Pull request by sorpaas**
This adds the `DefaultVote` trait to `pallet-collective`, allowing the default voting strategy to be customized. Two struct implementations are provided directly:

* `PrimeDefaultVote`: The current behavior -- use the prime vote as default vote.
* `MoreThanMajorityThenPrimeDefaultVote`: Check if raw yes votes are over majority of the total member count. If true, set default vote to Aye. Otherwise, use the prime default vote.

polkadot companion: https://github.com/paritytech/polkadot/pull/1661

### CODE REVIEW

**Review by shawntabrizi — APPROVED** [review]
lgtm.

**Comment by thiolliere** [comment]
the documentation for prime should be update:
```
 		/// The member who provides the default vote for any other members that do not vote before
		/// the timeout. If None, then no member has that privilege.
		pub Prime get(fn prime): Option<T::AccountId>;
```

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -71,6 +71,52 @@ pub type ProposalIndex = u32;
...
+/// Default voting strategy when a member is inactive.
+pub trait DefaultVote {
+	/// Get the default voting strategy, given:
+	///
+	/// - Whether the prime member voted Aye.
+	/// - Raw number of yes votes.
+	/// - Raw number of no votes.
+	/// - Total number of member count.
+	fn default_vote(
+		prime_voted_aye: bool,
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:83]: why not providing the complete prime_vote information as `Option<bool>` ?

and then rename PrimeDefaultVote to PrimeOrFalseDefaultVote ?

Comment (sorpaas) [reply frame/collective/src/lib.rs]: Changed it to expose the whole `prime_vote`! A small grumble on the rename though -- I think `PrimeDefaultVote` already expresses the whole intention, and `PrimeOrFalseDefaultVote` is slightly redundant. If we rename it, we also need to rename `MoreThanMajorityThenPrimeDefaultVote` to `MoreThanMajorityThenPrimeOrFalseDefaultVote` which I think is too long.

**Comment by gavofyork** [comment]
missing a test. looks good otherwise.

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (thiolliere) [reply frame/collective/src/lib.rs]: Yes I agree

**Comment by sorpaas** [comment]
bot merge

**Review by thiolliere — APPROVED** [review]

**Comment by sorpaas** [comment]
bot merge

**Comment by redzsina** [comment]
We are done with our review for this PR - please refer to issues #17 and #18 in the internal tracker.

**Pull request merged**